test/program_test.go

```diff
@@ -0,0 +1,242 @@
+
+			// wait until the controller has seen the stack object and completed processing it
+			var s pulumiv1.Stack
+			Eventually(func() bool {
+				err := k8sClient.Get(context.TODO(), types.NamespacedName{Namespace: stack.Namespace, Name: stack.Name}, &s)
+				if err != nil {
+					return false
+				}
+				if s.Generation == 0 {
+					return false
+				}
+				return s.Status.ObservedGeneration == s.Generation
+			}, "20s", "1s").Should(BeTrue())
+
+			var c corev1.ConfigMap
+			Eventually(func() bool {
+				err := k8sClient.Get(context.TODO(), types.NamespacedName{Namespace: stack.Namespace, Name: "test-configmap"}, &c)
+				fmt.Println(c)
+				return err == nil
+			}, "20s", "1s").Should(BeTrue())
+
+			Expect(s.Status.LastUpdate.State).To(Equal(shared.SucceededStackStateMessage))
+			Expect(apimeta.IsStatusConditionTrue(s.Status.Conditions, pulumiv1.ReadyCondition)).To(BeTrue())
+
+			By("Deleting the Stack Pulumi API Secret")
+			Expect(k8sClient.Delete(context.TODO(), &c)).Should(Succeed())
+			Eventually(func() bool {
+				err := k8sClient.Get(context.TODO(), types.NamespacedName{Name: c.Name, Namespace: namespace}, &c)
+				return k8serrors.IsNotFound(err)
+			}, k8sOpTimeout, interval).Should(BeTrue())
+		})
+
+		It("should fail if given a non-existent program.", func() {
+			stack := pulumiv1.Stack{
+				Spec: shared.StackSpec{
+					Stack:   randString(),
+					Backend: fmt.Sprintf("file://%s", backendDir),
+					ProgramRef: &shared.ProgramReference{
+						Name: randString(),
+					},
+					EnvRefs: map[string]shared.ResourceRef{
+						"PULUMI_CONFIG_PASSPHRASE": shared.NewLiteralResourceRef("password"),
+					},
+				},
+			}
+			stack.Name = randString()
+			stack.Namespace = "default"
+
+			Expect(k8sClient.Create(context.TODO(), &stack)).To(Succeed())
+
+			// wait until the controller has seen the stack object and completed processing it
+			var s pulumiv1.Stack
+			Eventually(func() bool {
+				err := k8sClient.Get(context.TODO(), types.NamespacedName{Namespace: stack.Namespace, Name: stack.Name}, &s)
+				if err != nil {
+					return false
+				}
+				if s.Generation == 0 {
+					return false
+				}
+				return s.Status.ObservedGeneration == s.Generation
+			}, "20s", "1s").Should(BeTrue())
+
+			Expect(s.Status.LastUpdate.State).To(Equal(shared.FailedStackStateMessage))
+			Expect(apimeta.IsStatusConditionFalse(s.Status.Conditions, pulumiv1.ReadyCondition)).To(BeTrue())
+		})
+
+		It("should fail if given a syntactically correct but invalid program.", func() {
+			prog := programFromFile("./testdata/test-program-invalid.yaml")
+			Expect(k8sClient.Create(context.TODO(), &prog)).To(Succeed())
+
+			stack := pulumiv1.Stack{
+				Spec: shared.StackSpec{
+					Stack:   randString(),
+					Backend: fmt.Sprintf("file://%s", backendDir),
+					ProgramRef: &shared.ProgramReference{
+						Name: prog.Name,
+					},
+					EnvRefs: map[string]shared.ResourceRef{
+						"PULUMI_CONFIG_PASSPHRASE": shared.NewLiteralResourceRef("password"),
```

Comment:
I think you'll also need the kubeconfig to make sure the failure isn't related to lack of creds. Same for previous test case.